README.md
# Circular Pulse [![GitHub stars](https://img.shields.io/github/stars/ProfessorX737/circular_pulse.svg?style=social)](https://github.com/ProfessorX737/circular_pulse) [![Twitter Follow](https://img.shields.io/twitter/url/https/@ayushpgupta.svg?style=social)](https://twitter.com/ayushpgupta) ![GitHub last commit](https://img.shields.io/github/last-commit/ProfessorX737/circular_pulse.svg) [![Website shields.io](https://img.shields.io/website-up-down-green-red/http/shields.io.svg)](https://play.google.com/store/apps/details?id=com.coddu.flutterprofile)[![Open Source Love](https://badges.frapsoft.com/os/v2/open-source.svg?v=103)](https://github.com/ProfessorX737/circular_pulse)

This Flutter package provides a Circular Pulse Widget with cool background glowing animation. Forked from https://pub.dev/packages/avatar_glow put re-published for control/stability

# 💻 Installation
In the `dependencies:` section of your `pubspec.yaml`, add the following line:

[![Version](https://img.shields.io/pub/v/circular_pulse.svg)](https://pub.dartlang.org/packages/circular_pulse)

```yaml
dependencies:
  circular_pulse: <latest version>
```

To use the latest changes:

```yaml
  circular_pulse:
    git:
      url: https://github.com/ProfessorX737/circular_pulse
      ref: master
```

# ❔ Usage

### Import this class

```dart
import 'package:circular_pulse/circular_pulse.dart';
```

Usage is simple. Circular Pulse is a widget offering different customizable optional parameters with child displayed at its center.

### - Simple Implementation
```dart
CircularPulse(
 endRadius: 60.0,
 child: Material(     // Replace this child with your own
   elevation: 8.0,
   shape: CircleBorder(),
   child: CircleAvatar(
     backgroundColor: Colors.grey[100],
     child: Image.asset(
       'assets/images/dart.png',
       height: 50,
     ),
     radius: 30.0,
   ),
 ),
),
```

### - Full Implementation
```dart
CircularPulse(
 glowColor: Colors.blue,
 endRadius: 90.0,
 duration: Duration(milliseconds: 2000),
 repeat: true,
 showTwoGlows: true,
 repeatPauseDuration: Duration(milliseconds: 100),
 child: Material(     // Replace this child with your own
   elevation: 8.0,
   shape: CircleBorder(),
   child: CircleAvatar(
     backgroundColor: Colors.grey[100],
     child: Image.asset(
       'assets/images/flutter.png',
       height: 60,
     ),
     radius: 40.0,
   ),
 ),
),
```

# 👍 Contribution
1. Fork it
2. Create your feature branch (git checkout -b my-new-feature)
3. Commit your changes (git commit -m 'Add some feature')
4. Push to the branch (git push origin my-new-feature)
5. Create new Pull Request


CHANGELOG.md
## Version 2.0.2

- Forked from apgapg/avatar_glow
- Fixed bug with late initialisation of _repeatPauseTimer
- Fixex bug with missing leading underscore to `controller`



lib/circular_pulse.dart
import 'dart:async';
import 'package:flutter/material.dart';

class CircularPulse extends StatefulWidget {
  final Widget child;
  final double endRadius;
  final BoxShape shape;
  final Duration duration;
  final bool repeat;
  final bool animate;
  final Duration repeatPauseDuration;
  final Curve curve;
  final bool showTwoGlows;
  final Color glowColor;
  final Duration? startDelay;

  const CircularPulse({
    Key? key,
    required this.child,
    required this.endRadius,
    this.shape = BoxShape.circle,
    this.duration = const Duration(milliseconds: 2000),
    this.repeat = true,
    this.animate = true,
    this.repeatPauseDuration = const Duration(milliseconds: 100),
    this.curve = Curves.fastOutSlowIn,
    this.showTwoGlows = true,
    this.glowColor = Colors.white,
    this.startDelay,
  }) : super(key: key);

  @override
  CircularPulseState createState() => CircularPulseState();
}

class CircularPulseState extends State<CircularPulse>
    with SingleTickerProviderStateMixin {
  late final _controller = AnimationController(
    duration: widget.duration,
    vsync: this,
  );
  late final _curve = CurvedAnimation(
    parent: _controller,
    curve: widget.curve,
  );
  late final _smallDiscAnimation = Tween(
    begin: (widget.endRadius * 2) / 6,
    end: (widget.endRadius * 2) * (3 / 4),
  ).animate(_curve);
  late final _bigDiscAnimation = Tween(
    begin: 0.0,
    end: widget.endRadius * 2,
  ).animate(_curve);
  late final _alphaAnimation = Tween(
    begin: 0.30,
    end: 0.0,
  ).animate(_controller);

  Timer? _repeatPauseTimer;

  void _statusListener(AnimationStatus status) async {
    if (_controller.status == AnimationStatus.completed) {
      _repeatPauseTimer = Timer(widget.repeatPauseDuration, () {
        if (mounted && widget.repeat && widget.animate) {
          _controller.reset();
          _controller.forward();
        }
      });
    }
  }

  @override
  void initState() {
    super.initState();
    if (widget.animate) {
      _startAnimation();
    }
  }

  @override
  void didUpdateWidget(CircularPulse oldWidget) {
    if (widget.animate != oldWidget.animate) {
      if (widget.animate) {
        _startAnimation();
      } else {
        _stopAnimation();
      }
    }
    super.didUpdateWidget(oldWidget);
  }

  Future<void> _startAnimation() async {
    _controller.addStatusListener(_statusListener);
    if (widget.startDelay != null) {
      await Future.delayed(widget.startDelay!);
    }
    if (mounted) {
      _controller.reset();
      _controller.forward();
    }
  }

  Future<void> _stopAnimation() async {
    _controller.removeStatusListener(_statusListener);
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: _alphaAnimation,
      child: widget.child,
      builder: (_, widgetChild) {
        final decoration = BoxDecoration(
          shape: widget.shape,
          // If the user picks a curve that goes below 0 or above 1
          // this opacity will have unexpected effects without clamping
          color: widget.glowColor.withOpacity(
            _alphaAnimation.value.clamp(0.0, 1.0),
          ),
        );
        return SizedBox(
          height: widget.endRadius * 2,
          width: widget.endRadius * 2,
          child: Stack(
            alignment: Alignment.center,
            children: <Widget>[
              if (widget.animate)
                AnimatedBuilder(
                  animation: _bigDiscAnimation,
                  builder: (_, __) {
                    // If the user picks a curve that goes below 0,
                    // this will throw without clamping
                    final _size =
                        _bigDiscAnimation.value.clamp(0.0, double.infinity);

                    return Container(
                      height: _size,
                      width: _size,
                      decoration: decoration,
                    );
                  },
                )
              else
                const SizedBox(height: 0.0, width: 0.0),
              if (widget.animate && widget.showTwoGlows)
                AnimatedBuilder(
                  animation: _smallDiscAnimation,
                  builder: (_, __) {
                    final size =
                        _smallDiscAnimation.value.clamp(0.0, double.infinity);

                    return Container(
                      height: size,
                      width: size,
                      decoration: decoration,
                    );
                  },
                )
              else
                const SizedBox(height: 0.0, width: 0.0),
              widgetChild!,
            ],
          ),
        );
      },
    );
  }

  @override
  void dispose() {
    _repeatPauseTimer?.cancel();
    _controller.dispose();
    super.dispose();
  }
}


example/lib/main.dart
import 'dart:io';

import 'package:circular_pulse/circular_pulse.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

// Sets a platform override for desktop to avoid exceptions. See
// https://flutter.dev/desktop#target-platform-override for more info.
void _enablePlatformOverrideForDesktop() {
  if (!kIsWeb && (Platform.isMacOS || Platform.isWindows || Platform.isLinux)) {
    debugDefaultTargetPlatformOverride = TargetPlatform.fuchsia;
  }
}

void main() {
  _enablePlatformOverrideForDesktop();
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Avatar Glow Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  bool _animate = true;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("Circular Pulse by @ProfessorX737 originally by @apgapg"),
      ),
      body: Container(
        child: ListView(
          children: <Widget>[
            Container(
              width: double.infinity,
              color: Color(0xff686de0),
              child: CircularPulse(
                endRadius: 70,
                animate: _animate,
                child: Material(
                  elevation: 8.0,
                  shape: CircleBorder(),
                  child: CircleAvatar(
                    backgroundColor: Colors.grey[100],
                    child: Image.asset(
                      'assets/images/dart.png',
                      height: 50,
                    ),
                    radius: 30,
                  ),
                ),
              ),
            ),
            Container(
              width: double.infinity,
              color: Color(0xff686de0),
              child: CircularPulse(
                startDelay: Duration(milliseconds: 1000),
                glowColor: Colors.white,
                endRadius: 100.0,
                duration: Duration(milliseconds: 2000),
                repeat: true,
                showTwoGlows: true,
                repeatPauseDuration: Duration(milliseconds: 100),
                child: Material(
                  elevation: 8.0,
                  shape: CircleBorder(),
                  color: Colors.transparent,
                  child: CircleAvatar(
                    backgroundImage: AssetImage('assets/images/avatar.png'),
                    radius: 50.0,
                  ),
                ),
                shape: BoxShape.circle,
                animate: true,
                curve: Curves.fastOutSlowIn,
              ),
            ),
            SizedBox(
              height: 16.0,
            ),
            Row(
              mainAxisSize: MainAxisSize.min,
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: <Widget>[
                CircularPulse(
                  glowColor: Colors.blue,
                  endRadius: 90.0,
                  duration: Duration(milliseconds: 2000),
                  repeat: true,
                  showTwoGlows: true,
                  repeatPauseDuration: Duration(milliseconds: 100),
                  child: Material(
                    elevation: 8.0,
                    shape: CircleBorder(),
                    child: CircleAvatar(
                      backgroundColor: Colors.grey[100],
                      child: Image.asset(
                        'assets/images/flutter.png',
                        height: 60,
                      ),
                      radius: 40.0,
                    ),
                  ),
                ),
                CircularPulse(
                  glowColor: Colors.red,
                  endRadius: 90.0,
                  duration: Duration(milliseconds: 2000),
                  repeat: true,
                  showTwoGlows: true,
                  repeatPauseDuration: Duration(milliseconds: 100),
                  child: Material(
                    elevation: 8.0,
                    shape: CircleBorder(),
                    child: CircleAvatar(
                      backgroundColor: Colors.grey[100],
                      child: Image.asset(
                        'assets/images/love.png',
                        height: 60,
                      ),
                      radius: 40.0,
                    ),
                  ),
                ),
              ],
            ),
            CircularPulse(
              glowColor: Colors.cyan,
              endRadius: 60.0,
              duration: Duration(milliseconds: 2000),
              repeat: true,
              showTwoGlows: true,
              repeatPauseDuration: Duration(milliseconds: 100),
              child: Material(
                elevation: 8.0,
                shape: CircleBorder(),
                child: CircleAvatar(
                  backgroundColor: Colors.grey[100],
                  child: Image.asset(
                    'assets/images/dart.png',
                    height: 50,
                  ),
                  radius: 30.0,
                ),
              ),
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          setState(() {
            _animate = !_animate;
          });
        },
      ),
    );
  }
}
